File: nim/src/hit_record.nim
import vec3
import point3
import ray

type
  HitRecord* = object
    p*: Point3
    normal*: Vec3
    t*: float
    frontFace*: bool

func qualifyNormal*(self: HitRecord; ray: Ray; outwardNormal: Vec3): HitRecord =
  let frontFace = dot(ray.direction, outwardNormal) < 0.0;
  let normal = if frontFace:
    outwardNormal
  else:
    -outwardNormal
  result = HitRecord(p: self.p, normal: normal, t: self.t, frontFace: frontFace)

File: rust/src/hit_record.rs
use crate::material::Material;
use crate::point3::Point3;
use crate::ray::Ray;
use crate::vec3::Vec3;

pub struct HitRecord {
    pub p: Point3,
    pub normal: Vec3,
    pub material: Material,
    pub t: f32,
    pub front_face: bool,
}

impl HitRecord {
    pub fn qualified_hit_record(
        ray: &Ray,
        m: Material,
        p: Point3,
        t: f32,
        outward_normal: Vec3,
    ) -> HitRecord {
        let front_face = ray.direction.dot(outward_normal) < 0.0;
        let normal = if front_face {
            outward_normal
        } else {
            -outward_normal
        };
        HitRecord {
            p: p,
            normal: normal,
            material: m,
            t: t,
            front_face: front_face,
        }
    }
}
